Opportunistically match built-in elements in createElementNS

rniwa opened the pull request:
#### 0a209521829e185dec49e36f65f9c6bca28dfc2b
<pre>
Opportunistically match built-in elements in createElementNS
<a href="https://bugs.webkit.org/show_bug.cgi?id=263151">https://bugs.webkit.org/show_bug.cgi?id=263151</a>

Reviewed by Chris Dumez.

Look up the element factory before validating the qualified name.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createElementNS):
(WebCore::isValidNameASCIIWithoutColon):
(WebCore::Document::parseQualifiedName):

Canonical link: <a href="https://commits.webkit.org/269388@main">https://commits.webkit.org/269388@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff817f6d08348eeab0e79fd993d88888796058a8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22266 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22465 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23338 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24163 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20604 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22513 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26727 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22792 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21615 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22499 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22205 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19319 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25017 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19254 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20175 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26438 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20272 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20405 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24284 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20932 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17728 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20395 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24622 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2810 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20930 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (approved):
Nice. LGTM if the bots are happy.

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1404,6 +1404,21 @@ CustomElementNameValidationStatus Document::validateCustomElementName(const Atom
 
 ExceptionOr<Ref<Element>> Document::createElementNS(const AtomString& namespaceURI, const AtomString& qualifiedName)
 {
+    auto opportuniticallyMatchedBuiltinElement = ([&]() -> RefPtr<Element> {
```
Typo: `opportunitically`

rniwa committed:
Opportunistically match built-in elements in createElementNS
https://bugs.webkit.org/show_bug.cgi?id=263151

Reviewed by Chris Dumez.

Look up the element factory before validating the qualified name.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createElementNS):
(WebCore::isValidNameASCIIWithoutColon):
(WebCore::Document::parseQualifiedName):

Canonical link: https://commits.webkit.org/269388@main

rkirsling commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1404,6 +1404,21 @@ CustomElementNameValidationStatus Document::validateCustomElementName(const Atom
+        if (settings().mathMLEnabled() && namespaceURI == MathMLNames::mathmlNamespaceURI)
+            return MathMLElementFactory::createKnownElement(qualifiedName, *this, /* createdByParser */ false);
+        return nullptr;
+    })();
+#endif
```
This is a syntax error when !ENABLE(MATHML). Fixed in #19148.
